orttraining/orttraining/python/orttraining_python_module.cc

```diff
@@ -357,18 +359,20 @@ PYBIND11_MODULE(onnxruntime_pybind11_state, m) {
 
   m.def("get_version_string", []() -> std::string { return ORT_VERSION; });
 
-  m.def("clear_training_ep_instances", []() -> void {
-    ort_training_env->ClearExecutionProviderInstances();
-  },
-        "Clean the execution provider instances used in ort training module.");
+  m.def(
+      "clear_training_ep_instances", []() -> void {
+        GetTrainingEnv().ClearExecutionProviderInstances();
+      },
+      "Clean the execution provider instances used in ort training module.");
 
   // clean the ort training environment when python interpreter exit
   // otherwise the global var will be de-constrcut after user main.
   // the order of ort training environment deconstruction and cudart
   // deconstruction is not stable, which will lead to crash.
   auto atexit = py::module_::import("atexit");
   atexit.attr("register")(py::cpp_function([]() {
-    ort_training_env = nullptr;
+    std::cout << "OrtTraining pybind11 atexit() is called" << std::endl;
+    GetTrainingEnv().ClearExecutionProviderInstances();
 #ifdef ENABLE_EAGER_MODE
     ort_backends_manager_instance = nullptr;
```

Comment:
> ort_backends_manager_instance = nullptr; [](http://example.com/codeflow?start=3&length=41)

Do we also want this to make function static and let it destruct naturally?

Reply:
We should not use python's  "atexit" at all. It runs while ORT python objects are still live. For example, if you have a session or global EP, "GetTrainingEnv().ClearExecutionProviderInstances();" will be called before these objects are destroyed.

Reply:
In my experiments, all python objects, specifically I was interested in session objects, were GCed before the module was unloaded. I hesitated adding it here, but unloading them automatically caused problems on Windows.

Reply:
The python layer, in this case, directly constructs its own instances of providers, in this instance CUDA and caches them in a map that lives in a global object. 
That global object is a part of python module and is destroyed after the providers DLLs are unloaded bc it does not hold handles on it. 
`atexit()` does clear provider instances before provider DLLs are unloaded.
You can propose a better solution to implement, but it may be out of scope for this PR.

Reply:
So, any global variable that is created after this ort_training_env is fully initialized will be destroyed after this object too. If I understand correctly, the atexit function calls after python objects are GC-ed but before the C++ part is started to tear down. So at this moment we can expect all C++ global variables are still alive.

Reply:
Yes, correct. However, now it is being enforced.

Reply:
Only for the ExecutionProviderInstances. It does not count the global variables or thread local variables that EPs have.